mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,727 @@
+        if block.inner().hint() != BLOCK_HINT {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
+        }
+
+        let data_size;
+        {
+            let inner_block = block.inner_mut();
+            inner_block.set_should_not_overwrite(false);
+            data_size = inner_block.data_size();
+        }
+
+        let should_flush = self.should_flush();
+        save_block_header(self.mmap, &block, start, self.max_file_size, should_flush)?;
+
+        end += data_size;
+
+        self.pointers.read = end % self.max_file_size;
+
+        self.flush_state
+            .update(0, 0, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        Ok(())
+    }
+
+    fn set_waker(&mut self, waker: &Waker) {
+        self.waker = Some(waker.clone());
+    }
+
+    fn should_flush(&self) -> bool {
+        match self.flush_options {
+            FlushOptions::AfterEachWrite => true,
+            FlushOptions::AfterXWrites(xwrites) => self.flush_state.writes >= xwrites,
+            FlushOptions::AfterXBytes(xbytes) => self.flush_state.bytes_written >= xbytes,
+            FlushOptions::AfterXMilliseconds(xmillis_elapsed) => {
+                self.flush_state.millis_elapsed >= xmillis_elapsed
+            }
+            FlushOptions::Off => false,
+        }
+    }
+
+    fn wake_up_task(&mut self) {
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+    }
+}
+
+#[allow(dead_code)]
+impl Drop for RingBuffer<'_> {
+    fn drop(&mut self) {
+        let result = self.mmap.flush();
+        if let Some(err) = result.err() {
+            error!("Failed to flush {:?}", err);
+        }
+    }
+}
+
+#[allow(dead_code)]
+fn create_file(file_path: &Path) -> IOResult<File> {
+    OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(file_path)
+}
+
+#[allow(dead_code)]
+fn find_pointers_post_crash(mmap: &MmapMut, max_file_size: usize) -> StorageResult<FilePointers> {
+    let mut block: BlockHeaderWithHash;
+    let block_size = serialized_block_size()?;
+
+    let mut start = 0;
+    let mut end = start + block_size;
+    let mut read = 0;
+    let mut write = 0;
+    // First we need to find *a* block header to work with.
+    // Once we find one, if any, then we can skip more efficiently
+    // to others.
+    loop {
+        if end > max_file_size {
+            return Ok(FilePointers { write, read });
+        }
+
+        if let Ok(next_block) = load_block_header(mmap, start, block_size, max_file_size) {
+            if next_block.inner().hint() == BLOCK_HINT {
+                block = next_block
+            } else {
+                start += 1;
+                end += 1;
+                continue;
+            }
+        } else {
+            start += 1;
+            end += 1;
+            continue;
+        }
+
+        break;
```

Comment:
not needed?

Reply:
Actually, I misspoke, we do want to break after we have found the first block